### PR TITLE
feature voting

### PR DESCRIPTION
**Pull request by summer7xinting**
Implemented
 - cli: qnode-score (feature: qnode_score)
 - cli: feature-voting

### CODE REVIEW

File: tl/cli/feature-voting.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+def parser():
+    return {
+        'help': """
+        Score candidates using feature voting. List of features may be used: 
+        1. exact match top 1: using pre-computed embedding vectors, either from a file or from elasticsearch.
+        2. page rank 
+        3. qnode with smallest number
+        4. Weiss Waterman distance
+        5. Jaccard between description and row
+        6. column query (SemTab 2020 paper)
```
Comment (saggu) [inline tl/cli/feature-voting.py:17]: Where are these options in the command line interface ? In other words, how can user pick and choose between these options?

File: tl/cli/feature-voting.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+def add_arguments(parser):
+    # input file
+    parser.add_argument('input_file', nargs='?', type=argparse.FileType('r'), default=sys.stdin)
+
+    # Features used in voting
+    parser.add_argument(
+        '-c', '--input-column-names', action='store', dest='input_column_names',
+        default='qnode_score',
+        help="the column name of features to use in voting, separated by ','"
```
Comment (saggu) [inline tl/cli/feature-voting.py:30]: should add options here

File: tl/cli/qnode-score.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import sys
```
Comment (saggu) [inline tl/cli/qnode-score.py:1]: please rename this file to `smallest-qnode-number`

File: tl/features/qnode_score.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import pandas as pd
```
Comment (saggu) [inline tl/features/qnode_score.py:1]: rename this file to `smallest_qnode_number`

File: tl/features/string_similarity.py [hunk 5]
```diff
@@ -76,6 +76,7 @@ def get_similarity_score(self):
         for _, each_row in self.df.iterrows():
             for each_similarity_unit in self.similarity_units:
                 # the output column name, should be the type + config for this similarity calculation unit
+                # TODO: naming is weird
```
Comment (saggu) [inline tl/features/string_similarity.py:79]: what are your changes in this file?

**Review by saggu — CHANGES_REQUESTED** [review]
@summer7xinting please check the review comments

**Comment by saggu** [comment]
Also, for any new command, please add the relevant entry in the REAME.md.

And pull requests go into the `development` branch

**Comment by summer7xinting** [comment]
need to rebase on dev branch :(

**Comment by saggu** [comment]
> need to rebase on dev branch :(

no, just pull from `usc-isi-i2:development` and create a new pull request

**Comment by summer7xinting** [comment]
Hey @saggu I think Ke-Thia's changes (external_embedding) is not in development which my work (voting) relies on. I can cherry pick my changes on top of dev but that won't be too useful without the embedding computation code 😔

File: tl/features/string_similarity.py (hunk 5 above)
Comment (summer7xinting) [reply tl/features/string_similarity.py:79]: `string-similarity` actually doesn't have a -o option to name output column

Comment (summer7xinting) [reply tl/features/string_similarity.py]: I added output column option in my latest commit

File: tl/cli/feature-voting.py (hunk 2 above)
Comment (summer7xinting) [reply tl/cli/feature-voting.py:28]: user will not choose from given option, they can specify any numeric value column as feature

**Pull request merged**